### ISSUE TITLE
Can we delete metrics created in the client?

### ISSUE THREAD

**Issue by sbose2k17** [prometheus-net/prometheus-net#80]
I have a created a number of metrics like histograms and counters with different combination of labels. However, it looks that the scrape endpoint after a period of time is shooting up its latency due to the long list of metrics that's being scaped. Can we delete these metrics as a way to reset them at that point?

**Comment by sandersaares** [comment]
There are various possibilities to control the exported data set. Simplest of all would be to use the `Collector<T>.RemoveLabelled()` method - it is the exact opposite of the `.Labels()` method, removing the data with the specified set of label values. Does this satisfy your need?

**Comment by sandersaares** [comment]
The exporters (e.g. `MetricServer`) collect the data using an implementation of `ICollectorRegistry`. In prometheus-net there is one implementation, named `DefaultCollectorRegistry`. If no custom instance of this class is provided on startup, the `DefaultCollectorRegistry.Instance` singleton is used.

As an alternative to `.RemoveLabelled()`, you can simply clear all registrations by using `DefaultCollectorRegistry.Clear()`. This resets the data set to an empty state - on the next collection no metrics will be returned. You will need to re-register any collectors you want to use (using e.g. `Metrics.CreateCounter()` and its siblings).

If even this does not afford the desired level of flexibility, you can simply create your own implementation of `ICollectorRegistry` to manage data at an even finer level of control.

**Comment by sb2k17** [comment]
The issue we currently wanted to address is with how we can clear the number of metrics if it becomes very large. In such a case, we are experiencing severe performance degradation i.e. large response size ~64 Mb at each scrape. One of the factors to cause this problem has been emitting sparse metrics due to labels. However, we need a way to control the metrics we emit and as well as manage it.

**Comment by sandersaares** [comment]
I see. Did the above information help solve your need?

**Comment by shahabganji** [comment]
@sandersaares 

I want to clear all metrics created on my scrape end point whenever prometheus collcets the data to prevent the end point to become too large, I am using version `3.1.0`. where can I find `DefaultCollectorRegistry.Clear()`

**Comment by sandersaares** [comment]
Anything that happens with metrics should not depend on scrapes. That is an anti-pattern.

This library is designed to enable you to use a relatively static set of metrics - each module in your app exposes metrics related to it. There metrics are reasonably stable, maybe only depending on when a particular module is activated. You should not be re-creating metrics on a regular basis and this library is not designed to enable such a pattern. There is no direct equivalent to `DefaultCollectorRegistry.Clear()` in the v3 API.

What might come and go on a more regular basis is the different combinations of labels - it is perfectly natural that e.g. when explorer.exe finishes work, metrics for explorer.exe might go away. For this purpose, there is `Collector.RemoveLabelled(string[] labels)`. You can use this to ensure that labelled metrics with by-now-irrelevant labels do not consume memory.

Were you using Clear() in order to remove obsolete labels? Or do you really have obsolete metrics? I would like to understand your usage scenario better, as input for future API design choices.

**Comment by jwenz723** [comment]
Is this information no longer valid? I cannot find the `DefaultCollectorRegistry.Clear()` method anywhere in the code within this repo.

**Comment by sandersaares** [comment]
That is correct, `DefaultCollectorRegistry.Clear()` was removed in v3. The metrics registry is intended as an append-only set, although you can still remove obsolete collections of label values via `RemoveLabelled()`.

In general, the set of exported metrics is intended to be static - removing them at runtime is not an intended use case of this library. If you can describe in detail why you need to remove metrics, I may be able to suggest an alternative approach.

**Comment by jwenz723** [comment]
@sandersaares I am trying to collect a thread status (value between 0 and 5) for a fixed number of threads (100). I thought a gauge would work well for this. However, I also want to collect the name of the process that is using each thread at the point in time when collection occurs. I thought I could just add this information using a label on each gauge, like:

thread_status{thread="1",process="myfirstprocess"} 0
thread_status{thread="2",process="mysecondprocess"} 3
thread_status{thread="3",process="mythirdprocess"} 2

The problem with this approach is that on the next collection interval if a different process is occupying the thread then a new timeseries is created and now there are multiple timeseries for each thread, and the previous values which are still being reported are no longer true:

thread_status{thread="1",process="myfirstprocess"} 0
thread_status{thread="2",process="mysecondprocess"} 3
thread_status{thread="3",process="mythirdprocess"} 2
thread_status{thread="1",process="myfourthprocess"} 0
thread_status{thread="2",process="mythirdprocess"} 3
thread_status{thread="3",process="myfirstprocess"} 2

So I was thinking on each collection interval of just removing all instances of the thread_status metric so that at any point in time only currently valid instances are displayed on the /metrics endpoint.

I imagine there is probably a better way to do this, but this was the best I could come up with, would appreciate any suggestions.

**Comment by cstockton** [comment]
@jwenz723 Sounds like you have threads a process P is running on and the status of a thread T. One alternative is to use two distinct sets of P and T:
```
process_thread{name="myfirstprocess"} 0 # on exit the process sets itself to 0
process_thread{name="mythirdprocess"} 2
process_thread{name="myfourthprocess"} 3
thread_status{id="1"} 0 # assuming 0 means "idle" (no running P) 
thread_status{id="2"} 4
thread_status{id="3"} 5
```

**Comment by jwenz723** [comment]
> @jwenz723 Sounds like you have threads a process P is running on and the status of a thread T. One alternative is to use two distinct sets of P and T:
> 
> ```
> process_thread{name="myfirstprocess"} 0 # on exit the process sets itself to 0
> process_thread{name="mythirdprocess"} 2
> process_thread{name="myfourthprocess"} 3
> thread_status{id="1"} 0 # assuming 0 means "idle" (no running P) 
> thread_status{id="2"} 4
> thread_status{id="3"} 5
> ```

Thanks for the suggestion. That sounds like a good reasonable approach.

**Comment by sandersaares** [comment]
Note that there is a difference between removing metrics and removing obsolete label values. From what I understand, your use case is removing label values, which is perfectly possible. Just use threadstatus.RemoveLabelled("1", "myfirstprocess") when the myfirstprocess value becomes stale.

Yes, you have to manually track which processes you have reported in order to remove them.

**Comment by Sakuni** [comment]
I also face the same problem with stale values and I want to know, is it bad to do a clear() on the gauge every time before publishing new set of series? As it describes in client, clear() removes all children and any references to any children are invalidated. Does this word "children" is referred to series?

**Comment by cstockton** [comment]
@Sakuni I would avoid clearing all your metrics if possible, since it creates a potential race condition where a scrape observes the metric between the clear and next round of updates. In general I try to define all my labels up front, then set gauges to the state of the value. Even in cases where I can no longer make a positive assertion about the value (say a network call to get the metric is unavailable) - I just create one or more additional states to reflect the failure modes.

I've found over time that when my first thought is to remove some labels or clear a metric value there's always an alternative design that avoids those problems. When there seems to be no other option, for example it would create too many metrics if I left old labels there - it tells me I need to reduce the cardinality since absent values are still a distinct vector in the tsdb clearing them doesn't make them "go away" - at least that's my understanding.

That said if cardinality isn't an issue maybe it is easier to just clear and recalculate all the metrics then try to do a differential update. In that case the only thing to be careful of is a race condition between the updates and prometheus scrapes giving you partial metric observations, i.e.:
  `(app: clear) -> (prometheus: scrape) -> (app: finished_updates))`

I haven't used .net but in Go there is a Gather() interface which you can implement, it provides a place to inject synchronization, e.g. simple mutex or single-flight mechanism. Maybe you can do similar in .net.

**Comment by Sakuni** [comment]
@cstockton Thank you for your explanation.

**Comment by Percygu** [comment]
> There are various possibilities to control the exported data set. Simplest of all would be to use the `Collector<T>.RemoveLabelled()` method - it is the exact opposite of the `.Labels()` method, removing the data with the specified set of label values. Does this satisfy your ne

> 

I just want to delete the No longer use metric in client,how can i do this with go library
